### ISSUE TITLE
[BYT][NOCODEC] alsa-bat check-alsabat-headset-playback.sh FAIL

### ISSUE THREAD

**Issue by plbossart** [thesofproject/linux#2397]
tracking https://sof-ci.01.org/linuxpr/PR2396/build4392/devicetest/

No idea what this happens.

**Comment by mengdonglin** [comment]
@singalsu Can you reproduce this failure locally?

**Comment by mengdonglin** [comment]
@plbossart @singalsu This seems to be the same random issue as https://github.com/thesofproject/sof/issues/1988 that was closed without a fix. 
Xiuli has checked the dumped WAV by CI and cannot see obvious wave distortion but cannot pass alsabat check.

**Comment by xiulipan** [comment]
@aiChaoSONG WAV files and pic uploaded. Please try some other measure method to confirm the error in the wav.
[bat.wav.zip](https://github.com/thesofproject/linux/files/5148547/bat.wav.zip)

![image](https://user-images.githubusercontent.com/37825714/91687290-f451ac80-eb91-11ea-9a53-3939055924e1.png)

**Comment by aiChaoSONG** [comment]
@keqiaozhang @mengdonglin @xiulipan @plbossart 
**Amp spectrum (whole wave, amp normalized to 1.0)**
![Screenshot from 2020-08-31 18-27-33](https://user-images.githubusercontent.com/52959957/91711452-7fde3400-ebb8-11ea-9703-216f7017a06d.png)
**Amp spectrum(whole wave, local area)**
![Screenshot from 2020-08-31 18-27-57](https://user-images.githubusercontent.com/52959957/91711454-8076ca80-ebb8-11ea-9fdf-e4283920a86a.png)
**side lobe peaks in dB:  -27.356dB -23.496dB**
**Amp spectrum (exclude the first 20000 samples, ignore ramp, local area)**
![Screenshot from 2020-08-31 18-47-53](https://user-images.githubusercontent.com/52959957/91712395-3a226b00-ebba-11ea-9600-2fdbea52d779.png)

The ramp at the beginning of the recorded wave introduces higher side lobe, which cause alsabat failed.
we can:
- add threshold in alsabat, make it less sensitive
- try to avoid the ramp at the very beginning

**Comment by singalsu** [comment]
@aiChaoSONG It would be better to plot with dB y-axis, but I agree with your proposals. The ramp in the beginning impacts the spectra.

**Comment by plbossart** [comment]
@aiChaoSONG @singalsu can't we just skip the first part of the captured stream? if we are doing a frequency analysis it has to be done in steady-state, no?

**Comment by singalsu** [comment]
Yep exactly, the ramp part need to be excluded from analysis to get state. Also the FFT needs to be windowed if it's length is not synchronized to sine wave period. I'm usually using "kaiser(N, beta)" and using beta that gives suitable tradeoff between sidelobes attenuation and tone frequency bins spread. When applying window the level seen for sine wave is also impacted. Maybe start with beta = 10 for kaiser.

For playing with impact of windows and FFT length interactively I recommend to check this bat.wav file with sound editor called "snd" (from Ubuntu packages). Check from waveform the "f" checkbox on. Then go to transform options  (single transform, dB, log freq, min dB 140, type Kaiser, size 1024, beta 0 - 1.0 (correspond to 0 - 15 as seen in plot), etc. If you select instead of single transform the sonogram, you'll see the distortion components of the ramp.

**Comment by aiChaoSONG** [comment]
I also checked wave file recorded by other platform, mainly JSL, the side lobe peak is not that high.

**Comment by singalsu** [comment]
@aiChaoSONG I edited my response above. To me the spectra in the bat.wav attached here looks quite OK. It's not 24 bit high-end quality but pretty good. There's some DC output above the noise floor (-100 dB or so, so very small, we may miss proper rounding somewhere in the test system and/or firmware but it's a minor issue).

**Comment by singalsu** [comment]
@aiChaoSONG The side lobes appear due to no windowing or insufficient windowing and due to different lenght of FFT than sine wave period (or integer multiple). E.g. the kaiser window with large enough beta is needed for >= 16 bit applications like SOF is.

**Comment by aiChaoSONG** [comment]
@singalsu In this case, alsabat is helping us to do wave analysis, after check its source code, I don't think it uses any window function.

https://github.com/alsa-project/alsa-utils/tree/master/bat

**Comment by singalsu** [comment]
Good finding @aiChaoSONG ! Computing of kaiser windows is somewhat tricky. Not sure any open-source library provide the needed functions. Does alsabat by chance use a fixed FFT length so the window could be pre-computed into a data file?

**Comment by aiChaoSONG** [comment]
> Good finding @aiChaoSONG ! Computing of kaiser windows is somewhat tricky. Not sure any open-source library provide the needed functions. Does alsabat by chance use a fixed FFT length so the window could be pre-computed into a data file?

@singalsu Unfortunately, it doesn't,  the FFT size it used is calculated by truncating   `time_length * sample_rate` to 2^N, and time_length is given by user from command line.

**Comment by singalsu** [comment]
There would be need for besseli() function to compute https://octave.sourceforge.io/signal/function/kaiser.html .

**Comment by singalsu** [comment]
A blackman window might do, the equation and example is found in https://ccrma.stanford.edu/~jos/mdft/Use_Blackman_Window.html .

**Comment by aiChaoSONG** [comment]
@singalsu The alsabat tool is too sensitive,  Do you think THD+N can be used in this scenario?  If yes, I can add implementation in wavetool.py, and we can use that for the alsabat test case.

**Comment by singalsu** [comment]
Yep, the THD+N test can be used instead. Please check also that sine signal level (measured with similar dBFS RMS) matches expected. Then it does the same as the FFT based check.
